File: backend/models/schemas/regulatory.py
"""Pydantic schemas for RegulatoryDocuments, RiskItems, and CAPAs."""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field


# ── RegulatoryDocument ────────────────────────────────────────────────────────

class RegDocCreate(BaseModel):
    doc_type: str
    title: str = Field(max_length=500)
    doc_number: str | None = Field(None, max_length=100)
    revision: str = "1.0"
    related_standards: list[str] = []
    effective_date: date | None = None
    next_review_date: date | None = None


class RegDocUpdate(BaseModel):
    title: str | None = Field(None, max_length=500)
    doc_number: str | None = None
    revision: str | None = None
    status: str | None = None
    related_standards: list[str] | None = None
    effective_date: date | None = None
    next_review_date: date | None = None
    reviewed_at: datetime | None = None
    approved_at: datetime | None = None


class RegDocOut(BaseModel):
    id: uuid.UUID
    document_id: uuid.UUID | None
    doc_type: str
    doc_number: str | None
    title: str
    revision: str
    status: str
    related_standards: list[str]
    owner_id: uuid.UUID | None
    reviewed_at: datetime | None
    approved_at: datetime | None
    effective_date: date | None
    next_review_date: date | None
    supersedes_id: uuid.UUID | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── RiskItem ──────────────────────────────────────────────────────────────────

class RiskItemCreate(BaseModel):
    regulatory_doc_id: uuid.UUID | None = None
    hazard: str
    hazardous_situation: str
    harm: str
    probability_before: int | None = Field(None, ge=1, le=5)
    severity_before: int | None = Field(None, ge=1, le=5)
    mitigation_measures: str | None = None
    probability_after: int | None = Field(None, ge=1, le=5)
    severity_after: int | None = Field(None, ge=1, le=5)
    risk_acceptability: str | None = None


class RiskItemUpdate(BaseModel):
    hazard: str | None = None
    hazardous_situation: str | None = None
    harm: str | None = None
    probability_before: int | None = Field(None, ge=1, le=5)
    severity_before: int | None = Field(None, ge=1, le=5)
    mitigation_measures: str | None = None
    probability_after: int | None = Field(None, ge=1, le=5)
    severity_after: int | None = Field(None, ge=1, le=5)
    risk_acceptability: str | None = None


class RiskItemOut(BaseModel):
    id: uuid.UUID
    regulatory_doc_id: uuid.UUID | None
    hazard: str
    hazardous_situation: str
    harm: str
    probability_before: int | None
    severity_before: int | None
    risk_score_before: int | None
    mitigation_measures: str | None
    probability_after: int | None
    severity_after: int | None
    risk_score_after: int | None
    risk_acceptability: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── CAPA ──────────────────────────────────────────────────────────────────────

class CAPACreate(BaseModel):
    capa_number: str = Field(max_length=100)
    title: str = Field(max_length=500)
    description: str | None = None
    capa_type: str = "capa"
    root_cause: str | None = None
    corrective_action: str | None = None
    preventive_action: str | None = None
    due_date: datetime | None = None
    owner_id: uuid.UUID | None = None


class CAPAUpdate(BaseModel):
    title: str | None = Field(None, max_length=500)
    description: str | None = None
    capa_type: str | None = None
    root_cause: str | None = None
    corrective_action: str | None = None
    preventive_action: str | None = None
    status: str | None = None
    due_date: datetime | None = None
    owner_id: uuid.UUID | None = None
    closed_at: datetime | None = None


class CAPAOut(BaseModel):
    id: uuid.UUID
    capa_number: str
    title: str
    description: str | None
    capa_type: str
    root_cause: str | None
    corrective_action: str | None
    preventive_action: str | None
    status: str
    owner_id: uuid.UUID | None
    due_date: datetime | None
    closed_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/regulatory.ts
// TypeScript types for Regulatory Documents, Risk Items, and CAPAs

export interface RegDoc {
  id: string;
  document_id: string | null;
  doc_type: string;
  doc_number: string | null;
  title: string;
  revision: string;
  status: string;
  related_standards: string[];
  owner_id: string | null;
  reviewed_at: string | null;
  approved_at: string | null;
  effective_date: string | null;
  next_review_date: string | null;
  supersedes_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface RegDocCreate {
  doc_type: string;
  title: string;
  doc_number?: string;
  revision?: string;
  related_standards?: string[];
  effective_date?: string;
  next_review_date?: string;
}

export interface RegDocUpdate {
  title?: string;
  doc_number?: string;
  revision?: string;
  status?: string;
  related_standards?: string[];
  effective_date?: string | null;
  next_review_date?: string | null;
}

export interface AIDraftResult {
  doc_id: string;
  content: string;
}

export interface RiskItem {
  id: string;
  regulatory_doc_id: string | null;
  hazard: string;
  hazardous_situation: string;
  harm: string;
  probability_before: number | null;
  severity_before: number | null;
  risk_score_before: number | null;
  mitigation_measures: string | null;
  probability_after: number | null;
  severity_after: number | null;
  risk_score_after: number | null;
  risk_acceptability: string | null;
  created_at: string;
  updated_at: string;
}

export interface CAPA {
  id: string;
  capa_number: string;
  title: string;
  description: string | null;
  capa_type: string;
  root_cause: string | null;
  corrective_action: string | null;
  preventive_action: string | null;
  status: string;
  owner_id: string | null;
  due_date: string | null;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface RiskItemCreate {
  hazard: string;
  hazardous_situation: string;
  harm: string;
  probability_before?: number | null;
  severity_before?: number | null;
  mitigation_measures?: string | null;
  probability_after?: number | null;
  severity_after?: number | null;
  risk_acceptability?: string | null;
}

export interface CAPACreate {
  capa_number: string;
  title: string;
  description?: string;
  capa_type?: string;
  due_date?: string;
}

export interface Briefing {
  id: string;
  user_id: string;
  type: string;
  headline: string | null;
  priority_items: unknown[] | null;
  open_actions: unknown[] | null;
  upcoming_events: unknown[] | null;
  full_content: string | null;
  generated_for_date: string;
  created_at: string;
}
